kolibri/core/content/api.py

```diff
@@ -1310,6 +1315,79 @@ def consolidate(self, items, queryset):
         return sorted_items
 
 
+class ContentDownloadRequestViewset(ValuesViewset):
+    pagination_class = OptionalPageNumberPagination
+
+    queryset = ContentDownloadRequest.objects.all()
+
+    field_map = {
+        "source_id": "self.request.user.id",
+    }
+
+    values = (
+        "id",
+        "requested_at",
+        "reason",
+        "contentnode_id",
+        "metadata",
+        "status",
+    )
+
+    def filter_queryset(self, queryset):
+        return queryset.filter(source_id=self.request.user.id)
+
+    def annotate_queryset(self, queryset):
+        return queryset.annotate(
+            has_removal=Exists(
+                ContentRemovalRequest.objects.filter(
+                    source_model=OuterRef("source_model"),
+                    source_id=OuterRef("source_id"),
+                    contentnode_id=OuterRef("contentnode_id"),
+                    requested_at__gte=OuterRef("requested_at"),
+                )
+            )
+        ).filter(has_removal=False)
+
+    def create(self, request, *args, **kwargs):
+        # if there is an existing deletion request, delete the deletion request
+        deletion_request = ContentRemovalRequest.objects.filter(
+            contentnode_id=self.request["contentnode_id"],
+            source_id=self.request.user.id,
+        )
+
+        if deletion_request:
+            ContentRemovalRequest.objects.filter(
+                contentnode_id=self.request["contentnode_id"],
+                source_id=self.request.user.id,
+            ).delete()
+        else:
+            try:
+                request = ContentDownloadRequest.build_for_user(self, *args)
+                # this is definitely not right but I don't know what would be? but I need to put that somewhere, I think
+                # i.e.
+                metadata = ContentNode.objects.filter(
+                    pk=OuterRef("contentnode_id")
+                ).metadata
+                ContentRequest.metadata = metadata
```

Comment:
So my assumption was that for the use case of a user request a download, we wouldn't immediately have the `ContentNode` metadata imported yet, but we would have some basic information for rendering it the UI. So I figured that until the metadata is imported, we could store that data in the `metadata` field of the new `ContentDownloadRequest`, like a cache, until `ContentNode` exists. So the create request would pass that along when creating the request.

In a complimentary manner, I figured the list response of this viewset could pull from either `metadata` or `ContentNode` so that it immediately displays appropriately in the UI.

Would that work with the frontend implementation of 'My Downloads'? Thoughts @rtibbles ?

Reply:
Yes, the current frontend posts the metadata to the backend, so this can be retrieved from the POST request rather than loading from the DB (which may well not exist locally yet).